src/services/auth.service.ts

```diff
@@ -1,41 +1,142 @@
+      if (!passwordValidate)
+        throw new Error('Password must be at least 8 characters long and contain a special character');
 
-    generateTokens(userId: string){
-        const accessToken = jwt.sign({sub: userId}, ACCESS_SECRET, {expiresIn: '15m'});
-        const refreshToken = jwt.sign({sub: userId}, REFRESh_SECRET, {expiresIn: '7d'});
-        return {accessToken, refreshToken};
+      const hashedPassword = await bcrypt.hash(password, 10);
+      let userRole = await this.roleRepo.findOne({ where: { name: 'USER' } });
+      if (!userRole) {
+        userRole = this.roleRepo.create({
+          name: 'USER',
+          description: 'Default role for new users',
+        });
+        await this.roleRepo.save(userRole);
+      }
+
+      const newUser = new User();
+      newUser.name = name;
+      newUser.email = email;
+      newUser.password = hashedPassword;
+      newUser.role = [userRole];
+      newUser.isActive = false;
+
+      const savedUser = await this.userRepo.save(newUser);
+
+      const token = jwt.sign({ userId: savedUser.id }, VERIFY_SECRET, { expiresIn: '1d' });
+      const activationLink = `${process.env.BASE_URL}/auth/verify?token=${token}`;
+
+      try {
+        const mailService = new MailService();
+        await mailService.sendActivationEmail(email, activationLink);
+      } catch (mailError) {
+        console.error('Email send failed, rolling back user:', mailError);
+        await this.userRepo.remove(savedUser);
+        throw new Error('Cannot send activation email');
+      }
+      delete savedUser.password;
+      return savedUser;
+    } catch (err) {
+      console.error('Register error:', err);
+      return null;
     }
+  }
 
-    async saveRefreshToken(user: User, token: string){
-        user.refreshToken = token;
-        await this.userRepo.save(user);
+  async verifyAccount(token: string) {
+    if (!token) {
+      throw new HttpException(400, 'TOKEN_MISSING', 'Missing activation token');
     }
 
-    async verifyRefreshToken(token: string){
-        try{
-            const payload = jwt.verify(token, REFRESh_SECRET) as any;
-            const user = await this.userRepo.findOne({where: {id: payload.sub}});
-            if(!user || user.refreshToken !== token) return null;
-            return user;
-        }catch{
-            return null;
-        }
+    try {
+      // Giải mã token
+      const payload = jwt.verify(token, VERIFY_SECRET) as { userId: string };
+
+      const user = await this.userRepo.findOne({ where: { id: payload.userId } });
+      if (!user) {
+        throw new HttpException(404, 'USER_NOT_FOUND', 'User not found');
+      }
+
+      // Kiểm tra nếu đã kích hoạt
+      if (user.isActive) {
+        return { alreadyActive: true };
+      }
+
+      // Kích hoạt tài khoản
+      user.isActive = true;
+      await this.userRepo.save(user);
+
+      return { alreadyActive: false };
+    } catch (err: any) {
+      if (err.name === 'TokenExpiredError') {
```

Comment:
Recommendation: In an if statement, using numeric comparison is faster than string comparison.